Missing plugins are being reported as uncaught exceptions

Reference: https://app.bugsnag.com/netlify/netlify-build/errors/5ea7d7bded7cfb00197383cd
Other: https://app.bugsnag.com/netlify/netlify-build/errors/5ea9c6d063104a0017720495

The following error is being reported as an uncaught exception:

```
Cannot find module 'netlify-plugin-gatsby-cache' from '/opt/build/repo'
```

It is coming from the `resolve` library. In theory, `resolve` errors should propagate and be handled. This might be an error in the `resolve` library itself.

--- a/packages/build/package.json
+++ b/packages/build/package.json
@@ -60,7 +60,7 @@
     "pretty-ms": "^6.0.1",
     "read-pkg-up": "^7.0.1",
     "readdirp": "^3.4.0",
-    "resolve": "^1.15.1",
+    "resolve": "^2.0.0-next.1",
     "semver": "^7.1.3",
     "string-width": "^4.2.0",
     "strip-ansi": "^6.0.0",

--- a/packages/build/src/install/missing.js
+++ b/packages/build/src/install/missing.js
@@ -1,13 +1,9 @@
-const { promisify } = require('util')
-
 const pFilter = require('p-filter')
-const resolve = require('resolve')
 
 const { logInstallMissingPlugins } = require('../log/main')
+const { resolveLocation } = require('../utils/resolve')
 
 const { addDependencies } = require('./main')
-
-const pResolve = promisify(resolve)
 
 // Automatically install plugins if not installed already
 const installMissingPlugins = async function({ pluginsOptions, buildDir, mode }) {
@@ -27,7 +23,7 @@ const isMissingPlugin = async function({ location, core }, buildDir) {
   }
 
   try {
-    await pResolve(location, { basedir: buildDir })
+    await resolveLocation(location, buildDir)
     return false
   } catch (error) {
     return true

--- a/packages/build/src/plugins/options.js
+++ b/packages/build/src/plugins/options.js
@@ -1,16 +1,12 @@
 const { dirname } = require('path')
-const { promisify } = require('util')
-
-const resolve = require('resolve')
 
 const corePackageJson = require('../../package.json')
 const { installMissingPlugins } = require('../install/missing')
 const { CORE_PLUGINS } = require('../plugins_core/main')
+const { resolveLocation } = require('../utils/resolve')
 
 const { useManifest } = require('./manifest/main')
 const { getPackageJson } = require('./package')
-
-const pResolve = promisify(resolve)
 
 // Load plugin options (specified by user in `config.plugins`)
 const getPluginsOptions = async function({ netlifyConfig: { plugins }, buildDir, constants: { FUNCTIONS_SRC }, mode }) {
@@ -30,7 +26,7 @@ const normalizePluginOptions = function({ package, location = package, core = fa
 // Retrieve plugin's main file path.
 // Then load plugin's `package.json` and `manifest.yml`.
 const loadPluginFiles = async function({ pluginOptions, pluginOptions: { location }, buildDir }) {
-  const pluginPath = await pResolve(location, { basedir: buildDir })
+  const pluginPath = await resolveLocation(location, buildDir)
   const pluginDir = dirname(pluginPath)
   const { packageDir, packageJson } = await getPackageJson({ pluginDir })
   const { manifest, inputs: inputsA } = await useManifest(pluginOptions, { pluginDir, packageDir, packageJson })

--- a/packages/build/src/utils/resolve.js
+++ b/packages/build/src/utils/resolve.js
@@ -1,0 +1,19 @@
+const resolve = require('resolve')
+
+// Like `require.resolve()` but works with a custom base directory.
+// We need to use `new Promise()` due to a bug with `utils.promisify()` on
+// `resolve`:
+//   https://github.com/browserify/resolve/issues/151#issuecomment-368210310
+const resolveLocation = function(location, basedir) {
+  return new Promise((success, reject) => {
+    resolve(location, { basedir }, (error, resolvedLocation) => {
+      if (error) {
+        return reject(error)
+      }
+
+      success(resolvedLocation)
+    })
+  })
+}
+
+module.exports = { resolveLocation }

